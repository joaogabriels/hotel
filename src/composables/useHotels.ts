import { ref, computed, watch } from 'vue'
import { useRoute } from 'vue-router'
import { hotelService } from '../services/hotelService'
import type { Hotel } from '../components/models'

export function useHotels() {
  const route = useRoute()
  const hotels = ref<Hotel[]>([])
  const isFetching = ref(false)
  const totalPages = ref(0)
  const currentPage = ref(1)

  const hasNextPage = computed(() => currentPage.value < totalPages.value)

  const fetchHotels = async () => {
    isFetching.value = true

    try {
      const response = await hotelService.getHotels({
        location: route.query.location as string,
        name: route.query.hotelName as string,
        sortBy: route.query.sortBy as 'recommended' | 'best_rated',
        page: Number(route.query.page || 1),
        perPage: 10,
      })

      hotels.value = response.data
      totalPages.value = response.totalPages
      currentPage.value = response.page
    } catch (error) {
      console.error('Error fetching hotels:', error)
      throw error
    } finally {
      isFetching.value = false
    }
  }

  const fetchNextPage = async () => {
    if (!hasNextPage.value || isFetching.value) return

    await fetchHotels()
  }

  watch(
    () => route.query,
    async () => {
      await fetchHotels()
    },
    { immediate: true },
  )

  return {
    hotels,
    isFetching,
    hasNextPage,
    fetchNextPage,
    currentPage,
    totalPages,
  }
}
