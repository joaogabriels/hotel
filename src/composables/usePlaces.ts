import { ref } from 'vue'

import type { Place } from '../components/models'
import { placeService } from '../services/placeService'

export function usePlaces() {
  const places = ref<Place[]>([])
  const isLoading = ref(false)
  const error = ref<Error | null>(null)

  const fetchPlaces = async () => {
    isLoading.value = true
    error.value = null

    try {
      places.value = await placeService.getPlaces()
    } catch (err) {
      error.value = err as Error
      console.error('Error loading places:', err)
    } finally {
      isLoading.value = false
    }
  }

  const filterPlaces = (searchTerm: string) => {
    if (!searchTerm) return places.value

    const needle = searchTerm.toLowerCase()
    return places.value.filter(
      (place) =>
        place.name.toLowerCase().includes(needle) ||
        place.state.name.toLowerCase().includes(needle),
    )
  }

  return {
    places,
    isLoading,
    error,
    fetchPlaces,
    filterPlaces,
  }
}
