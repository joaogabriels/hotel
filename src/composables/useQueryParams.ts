import { useRoute, useRouter } from 'vue-router'

interface QueryParams {
  location?: string
  hotelName?: string
  sortBy?: string
}

export function useQueryParams() {
  const router = useRouter()
  const route = useRoute()

  const updateQueryParams = async (params: Partial<QueryParams>) => {
    const newQuery = {
      ...route.query,
      ...params,
    }

    Object.keys(newQuery).forEach((key) => {
      if (
        newQuery[key as keyof QueryParams] === undefined ||
        newQuery[key as keyof QueryParams] === null ||
        newQuery[key as keyof QueryParams] === ''
      ) {
        delete newQuery[key as keyof QueryParams]
      }
    })

    await router.push({ query: newQuery })
  }

  return {
    updateQueryParams,
  }
}
