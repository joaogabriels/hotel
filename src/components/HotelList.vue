<template>
  <div class="column q-gutter-y-md">
    <template v-if="loading && !hotels.length">
      <q-skeleton v-for="n in 3" :key="`${n}-skeleton`" class="hotel-card-skeleton" height="320px" square />
    </template>

    <template v-else>
      <q-infinite-scroll @load="onLoad" :offset="250" ref="infiniteScrollRef">
        <template v-if="hotels.length">
          <div class="column q-gutter-y-md">
            <HotelCard v-for="hotel in hotels" :key="hotel.id" :hotel="hotel" @select="$emit('select', hotel)" />
          </div>
        </template>

        <template v-else>
          <div class="column items-center q-pa-xl">
            <q-icon name="search_off" size="4rem" color="grey-6" />
            <p class="text-h6 text-grey-6">Nenhum hotel encontrado</p>
          </div>
        </template>

        <template #loading>
          <div class="row justify-center q-my-md">
            <q-spinner-dots color="primary" size="40px" />
          </div>
        </template>

      </q-infinite-scroll>

      <div class="row justify-center" v-if="!hasNextPage && hotels.length">
        <p class="text-grey-6 text-subtitle2 text-weight-regular">Não existem mais hotéis para serem exibidos :(</p>
      </div>
    </template>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue'
import type { QInfiniteScroll } from 'quasar'
import HotelCard from './HotelCard.vue'
import { useHotels } from '../composables/useHotels'
import type { Hotel } from './models'

const infiniteScrollRef = ref<QInfiniteScroll>()

const {
  hotels,
  isFetching: loading,
  hasNextPage,
  fetchNextPage,
} = useHotels()

const onLoad = async (index: number, done: () => void) => {
  await fetchNextPage()
  done()
}

defineEmits<{
  select: [hotel: Hotel]
}>()
</script>

<style lang="scss" scoped>
.hotel-card-skeleton {
  border-radius: 12px;
}
</style>