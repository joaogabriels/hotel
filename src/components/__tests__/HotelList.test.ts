import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import type { Ref, ComputedRef } from 'vue'

import { mockHotel } from './mocks'
import * as useHotelsModule from '../../composables/useHotels'
import HotelList from '../HotelList.vue'
import type { Hotel } from '../models'

installQuasarPlugin()

describe('HotelList', () => {
  it('matches snapshot', () => {
    vi.spyOn(useHotelsModule, 'useHotels').mockReturnValue({
      hotels: [mockHotel] as unknown as Ref<Hotel[]>,
      isFetching: false as unknown as Ref<boolean>,
      hasNextPage: true as unknown as ComputedRef<boolean>,
      fetchNextPage: vi.fn(),
      currentPage: 1 as unknown as Ref<number>,
      totalPages: 2 as unknown as Ref<number>,
      reset: vi.fn(),
    })

    const wrapper = mount(HotelList)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('triggers next page fetch', async () => {
    const fetchNextPage = vi.fn()
    vi.spyOn(useHotelsModule, 'useHotels').mockReturnValue({
      hotels: [mockHotel] as unknown as Ref<Hotel[]>,
      isFetching: false as unknown as Ref<boolean>,
      hasNextPage: true as unknown as ComputedRef<boolean>,
      fetchNextPage,
      currentPage: 1 as unknown as Ref<number>,
      totalPages: 2 as unknown as Ref<number>,
      reset: vi.fn(),
    })

    const wrapper = mount(HotelList)
    const infiniteScroll = wrapper.find('.q-infinite-scroll')

    await infiniteScroll?.trigger('scroll')

    expect(fetchNextPage).toHaveBeenCalled()
  })

  it('shows no more to load message when hasNextPage is false', () => {
    vi.spyOn(useHotelsModule, 'useHotels').mockReturnValue({
      hotels: [mockHotel] as unknown as Ref<Hotel[]>,
      isFetching: false as unknown as Ref<boolean>,
      hasNextPage: false as unknown as ComputedRef<boolean>,
      fetchNextPage: vi.fn(),
      currentPage: 2 as unknown as Ref<number>,
      totalPages: 2 as unknown as Ref<number>,
      reset: vi.fn(),
    })

    const wrapper = mount(HotelList)
    const noMoreMessage =
      wrapper.findAll('.text-grey-6')[wrapper.findAll('.text-grey-6').length - 1]

    expect(noMoreMessage?.exists()).toBe(true)
    expect(noMoreMessage?.text()).toBe('Não existem mais hotéis para serem exibidos :(')
  })
})
