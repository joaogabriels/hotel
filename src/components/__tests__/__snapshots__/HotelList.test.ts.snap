// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelList > matches snapshot 1`] = `
"<div data-v-b496c722=\\"\\" class=\\"column q-gutter-y-md\\">
  <div data-v-b496c722=\\"\\" class=\\"q-infinite-scroll\\">
    <div data-v-b496c722=\\"\\" class=\\"column q-gutter-y-md\\">
      <div data-v-b496c722=\\"\\" class=\\"q-card q-card--flat no-shadow hotel-card\\">
        <div class=\\"row full-width\\">
          <div class=\\"col-12 col-md-4\\">
            <div data-v-f3901ebe=\\"\\" class=\\"q-carousel q-panel-parent q-carousel--without-padding hotel-carousel\\" style=\\"height: 320px;\\">
              <div class=\\"q-carousel__slides-container\\">
                <transition-stub appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
                  <div class=\\"q-panel scroll\\" style=\\"--q-transition-duration: 300ms;\\" role=\\"tabpanel\\">
                    <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__slide\\"><img data-v-f3901ebe=\\"\\" src=\\"https://fakeimg.pl/600x400\\" class=\\"full-width full-height fit\\"></div>
                  </div>
                </transition-stub>
              </div>
              <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-left carousel-nav-btn\\" style=\\"margin: 160px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_back</i></span></button></div>
              <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-right carousel-nav-btn\\" style=\\"margin: 160px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_forward</i></span></button></div>
              <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-bottom slide-counter-container\\" style=\\"margin: 18px;\\">
                <div data-v-f3901ebe=\\"\\" class=\\"slide-counter\\">1/3</div>
              </div>
            </div>
          </div>
          <div class=\\"col-12 col-md-5\\">
            <div class=\\"q-card__section q-card__section--vert\\">
              <h2 class=\\"text-h5 q-ma-none text-grey-9\\">Hotel de Teste</h2>
              <h3 class=\\"text-subtitle1 text-grey-6 q-ma-none text-weight-regular\\">Av. Antônio Abrahão Caram, 1001, São José, Belo Horizonte, MG, Brasil, 31275-000</h3>
              <div class=\\"row items-center q-gutter-x-md q-mt-md\\">
                <div data-v-378063dc=\\"\\" class=\\"row items-start q-gutter-x-xs rating-section\\"><span data-v-378063dc=\\"\\">5</span>
                  <div data-v-378063dc=\\"\\" class=\\"q-rating row inline items-center q-rating--non-editable text-orange-4 rating\\" style=\\"font-size: 18px;\\" role=\\"radiogroup\\" aria-readonly=\\"true\\">
                    <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 1\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                    <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 2\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                    <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 3\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                    <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 4\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                    <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"true\\" aria-label=\\"grade 5\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                  </div>
                </div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">help_outline</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">restaurant_menu</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">store</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">hot_tub</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">medical_services</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">local_library</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">golf_course</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">surfing</i></div>
                <div data-v-accd225c=\\"\\" class=\\"row items-center amenity-item text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">child_care</i></div>
              </div>
              <div class=\\"q-mt-md\\">
                <div class=\\"q-chip row inline no-wrap items-center bg-grey-6 text-grey-1 q-chip--colored q-chip--dense no-shadow rounded-borders q-pa-md\\" readonly=\\"\\">
                  <div class=\\"q-chip__content col row no-wrap items-center q-anchor--skip\\">
                    <div class=\\"ellipsis\\">Não Reembolsável</div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class=\\"col-12 col-md-3 price-section\\">
            <div class=\\"q-card__section q-card__section--vert column full-height\\">
              <div class=\\"col-auto q-mb-md q-mb-md-none\\">
                <p class=\\"text-caption q-ma-none text-weight-regular text-grey-6\\">A partir de</p>
                <p class=\\"text-caption q-mb-sm\\"> R$ <span class=\\"text-h5\\">0,00</span></p>
                <p class=\\"text-caption q-ma-none text-weight-regular text-grey-6\\">Impostos inclusos</p>
              </div>
              <div class=\\"col full-width flex justify-start items-start\\"><button class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><span class=\\"block\\">Selecionar</span></span></button></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class=\\"q-infinite-scroll__loading invisible\\">
      <div data-v-b496c722=\\"\\" class=\\"row justify-center q-my-md\\"><svg data-v-b496c722=\\"\\" class=\\"q-spinner text-primary\\" fill=\\"currentColor\\" width=\\"40px\\" height=\\"40px\\" viewBox=\\"0 0 120 30\\" xmlns=\\"http://www.w3.org/2000/svg\\">
          <circle cx=\\"15\\" cy=\\"15\\" r=\\"15\\">
            <animate attributeName=\\"r\\" from=\\"15\\" to=\\"15\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"15;9;15\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate>
            <animate attributeName=\\"fill-opacity\\" from=\\"1\\" to=\\"1\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"1;.5;1\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate>
          </circle>
          <circle cx=\\"60\\" cy=\\"15\\" r=\\"9\\" fill-opacity=\\".3\\">
            <animate attributeName=\\"r\\" from=\\"9\\" to=\\"9\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"9;15;9\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate>
            <animate attributeName=\\"fill-opacity\\" from=\\".5\\" to=\\".5\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\".5;1;.5\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate>
          </circle>
          <circle cx=\\"105\\" cy=\\"15\\" r=\\"15\\">
            <animate attributeName=\\"r\\" from=\\"15\\" to=\\"15\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"15;9;15\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate>
            <animate attributeName=\\"fill-opacity\\" from=\\"1\\" to=\\"1\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"1;.5;1\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate>
          </circle>
        </svg></div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
